lib/spack/spack/test/cmd/env.py

```diff
@@ -1110,7 +1113,7 @@ def noop(*args):
         # according to the DAG hash (since build deps are excluded from
         # comparison by default). Although the dag hashes are equal, the specs
         # are not considered equal because they compare build deps.
-        assert x_concretized['y'].dag_hash() == y_concretized.dag_hash()
+        assert x_concretized['y'].runtime_hash() == y_concretized.runtime_hash()
```

Comment:
Does this test make sense anymore? Aren't we indexing things only by full-hash? If I understand correctly what happens now is that there will be to different `y` in the environment, or am I wrong?

Reply:
Ok, it's possible I'm missing something obvious, but to me it doesn't look like this test is installing anything, just working with an environment (add specs, concretize, and write).  So I don't get how the installation prefix comes into play.  Since the environment was previously keyed by the build hash, I think there should have been two different `y` present after concretization.

Just to check I ran the test with `develop` and put a breakpoint right before the final assert.  The lockfile on disk contains two different instances of 'y', which do have the same runtime hash (DAG hash at that point in time) but can both co-exist in the env due to different build hashes.

But after this PR, I agree we could actually install both of those `y` and not have any problems with overwriting.

I'm not sure if I think the test should be removed or maybe just renamed.

Reply:
I'll rephrase. My understanding before this PR:
1. There would have been 2 different `y` in the lockfile (since they have different `build_hash` and we keyed by `build_hash`)
2. Uniqueness of a spec was by `dag_hash` (i.e. by what we call now `runtime_hash`)
3. Therefore the `y` used as a link dependency by `x` was "the same" (i.e. same `dag_hash`) as the one installed as a standalone with a different `z` build dependency

The two `y` happen to be "the same" also on disk, but you're right that this is a detail that follows from the model above. The test in my understanding is there to ensure that we can treat this weirdness "correctly".

In the new model the `dag_hash` has the same meaning of the old `full_hash`, so we can distinguish between the two `y`. The scheme in the comment is therefore wrong imo - it should be `y1` and `y2` since they're not "the same" anymore. That leads me to think this test has lost its purpose with this PR and can be removed.

Reply:
I tweaked this a bit to do the same check but with `eq_node` instead of `runtime_hash()`.  We don't need to use the runtime hash here -- we can just check that the `y` *nodes* are the same but that the `z`'s and `y` *sub-dags* are not.

```python
        # Even though x chose a different 'z', the y it chooses should be identical
        # *aside* from the dependency on 'z'.  The dag_hash() will show the difference
        # in build dependencies.
        assert x_concretized['y'].eq_node(y_concretized)
        assert x_concretized['y'].dag_hash() != y_concretized.dag_hash()
```